src/prted/pmix/pmix_server_queries.c

```diff
@@ -123,7 +124,38 @@ static void _query(int sd, short args, void *cbdata)
         /* see if they provided any qualifiers */
         if (NULL != q->qualifiers && 0 < q->nqual) {
             for (n=0; n < q->nqual; n++) {
+                prte_output_verbose(2, prte_pmix_server_globals.output,
+                                    "%s qualifier key \"%s\" : value \"%s\"",
+                                    PRTE_NAME_PRINT(PRTE_PROC_MY_NAME),
+                                    q->qualifiers[n].key,
+                                    (q->qualifiers[n].value.type == PMIX_STRING ? q->qualifiers[n].value.data.string : "(not a string)"));
                 if (PMIX_CHECK_KEY(&q->qualifiers[n], PMIX_NSPACE)) {
+                    /* Never trust the namespace string that is provided.
+                     * First check to see if we know about this namespace. If
+                     * not then return an error. If so then continue on.
+                     * Note that the "PRTE_PMIX_CONVERT_NSPACE" function will create
+                     * a new prte_job_t structure an add it to the list.
+                     */
+                    /* Make sure the qualifier namespace exists */
+                    matched = 0;
+                    PRTE_HASH_TABLE_FOREACH(key, uint32, jdata, prte_job_data) {
+                        if (NULL != jdata &&
+                            PMIX_CHECK_NSPACE(q->qualifiers[n].value.data.string,
+                                              jdata->nspace)) {
+                            matched = 1;
+                            break;
+                        }
+                    }
+                    if (0 == matched) {
+                        prte_output_verbose(2, prte_pmix_server_globals.output,
+                                            "%s qualifier key \"%s\" : value \"%s\" is an unknown namespace",
+                                            PRTE_NAME_PRINT(PRTE_PROC_MY_NAME),
+                                            q->qualifiers[n].key,
+                                            q->qualifiers[n].value.data.string);
+                        ret = PMIX_ERR_BAD_PARAM;
```

Comment:
PMIX_ERR_BAD_PARM is probably a better error code choice but it's not listed as a return code for PMIx_Query_info in the standard, which is why I used PMIX_NOT_FOUND

Reply:
Yeah, it probably should be listed in the standard. I'll work on that.